### PR TITLE
Use base64 in JsonSerializer

### PR DESCRIPTION
**Pull request by shargon**
According to https://github.com/neo-project/neo/issues/1191#issuecomment-549672766

Optimization of https://github.com/neo-project/neo/pull/1197

### CODE REVIEW

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/Network/P2P/Payloads/Transaction.cs#L186

Should we update these places too?

**Comment by devhawk** [comment]
this is a major breaking change for NEO 3, right?

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/Ledger/ContractState.cs#L65

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/Network/P2P/Payloads/TransactionAttribute.cs#L33

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/Network/P2P/Payloads/Witness.cs#L47-L48

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/SmartContract/ContractParameter.cs#L150

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/SmartContract/ContractParametersContext.cs#L50

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/SmartContract/ContractParametersContext.cs#L56

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/SmartContract/ContractParametersContext.cs#L287

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/SmartContract/Manifest/ContractGroup.cs#L51

https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/Wallets/NEP6/NEP6Contract.cs#L27

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo/blob/1c84f7cf9f211f386107677ce28b26de26042677/neo/Network/RPC/Models/RpcNep5Balances.cs#L42

Also, I think this line should be:

```c#
json["asset_hash"] = AssetHash.ToString();
```

**Comment by shargon** [comment]
> this is a major breaking change for NEO 3, right?

Yes

File: neo/Helper.cs [hunk 1]
```diff
@@ -98,6 +98,16 @@ internal static string GetVersion(this Assembly assembly)
             return (string)attribute.ConstructorArguments[0].Value;
         }
 
+        public static string ToBase64String(this byte[] value)
```
Comment (erikzhang) [inline neo/Helper.cs]: We don't need this function, I think. Just use `Convert` directly.

File: neo/SmartContract/ContractParameter.cs [hunk 2]
```diff
@@ -103,7 +103,7 @@ public void SetValue(string text)
             switch (Type)
             {
                 case ContractParameterType.Signature:
-                    byte[] signature = text.HexToBytes();
+                    byte[] signature = text.Base64ToBytes();
```
Comment (erikzhang) [inline neo/SmartContract/ContractParameter.cs]: I think this is not JSON string.

File: neo/SmartContract/ContractParameter.cs [hunk 3]
```diff
@@ -120,7 +120,7 @@ public void SetValue(string text)
                     Value = UInt256.Parse(text);
                     break;
                 case ContractParameterType.ByteArray:
-                    Value = text.HexToBytes();
+                    Value = text.Base64ToBytes();
```
Comment (erikzhang) [inline neo/SmartContract/ContractParameter.cs]: Not a JSON string too.

File: neo/SmartContract/ContractParameter.cs [hunk 4]
```diff
@@ -197,7 +197,7 @@ private static string ToString(ContractParameter parameter, HashSet<ContractPara
                 case null:
                     return "(null)";
                 case byte[] data:
-                    return data.ToHexString();
+                    return data.ToBase64String();
```
Comment (erikzhang) [inline neo/SmartContract/ContractParameter.cs]: Not JSON.

File: neo/Helper.cs (hunk 1 above)
Comment (shargon) [reply neo/Helper.cs]: I added this because the argument, in order to prevent mistakes

Comment (erikzhang) [reply neo/Helper.cs]: I think the default value is `Base64FormattingOptions.None`.

**Comment by erikzhang** [comment]
Should we remove all the `Base64FormattingOptions.None`?

**Comment by shargon** [comment]
By default is `none`, so if yo prefer, i can do it

**Comment by erikzhang** [comment]
Please do it.

**Comment by cloud8little** [comment]
@shargon @erikzhang It's found that this change is  incompatible with previous preview1 json file, how do you handle it?

4111573617621_ pic_hd

**Comment by shargon** [comment]
We can convert to byte array if is hexadecimal too

**Pull request merged**